tasks/parameter/sp.yml

```diff
@@ -21,8 +21,8 @@
 
 - name: Set service principal parameter
   set_fact:
-      sp: "{{ sp | default({}) | combine({ item.key: vars[item.value] }) }}"
+      sp: "{{ sp | default({}) | combine({ item.key: lookup('vars', item.value) }) }}"
```

Comment:
`var[...]` yields the literal (untransformed) variable, using `lookup('vars', ...)` gives us the value the invoker intended

Reply:
I cannot see the difference. Could you please add an example of the output?.

To me doing something like:

```yaml
---
- name: Lookup Vars
  hosts: localhost
  gather_facts: false
  connection: local

  vars:
    test:
      - 1
      - 2
      - 3
      - "Text1"

  tasks:

    - name: Debug
      debug:
        msg: "{{ lookup('vars', 'test') }}"
      with_items: test

    - name: Debug
      debug:
        msg: "{{ vars[item] }}"
      with_items: test
```

Returns exactly the same output:

```bash
TASK [Debug] *********************************************************************************************************************************************************************************************
ok: [localhost] => (item=test) => {
    "msg": [
        1,
        2,
        3,
        "Text1"
    ]
}

TASK [Debug] *********************************************************************************************************************************************************************************************
ok: [localhost] => (item=test) => {
    "msg": [
        1,
        2,
        3,
        "Text1"
    ]
}
```

Reply:
@imjoseangel
The effect is indeed very subtle - you cannot use `debug` to show you the true var state since it will be transformed by debug itself. Try the following:

```yaml
---
- name: Lookup Vars
  hosts: localhost
  gather_facts: false
  connection: local

  vars:
    test_var: "test"
    service_principal: "{{ test_var }}"
    client_secret: "{{ test_var }}"

  tasks:

    - name: Set service principal parameter
      set_fact:
          sp: "{{ sp | default({}) | combine({ item.key: vars[item.value] }) }}"
      with_items:
          - {'key': 'client_id', 'value': 'service_principal' }
          - {'key': 'client_secret', 'value': 'client_secret' }
      when: service_principal is defined and vars[item.value] is defined

    - debug:
        var: service_principal
      when: "{{ lookup('vars', 'service_principal') == 'test' }}"

    - debug:
        var: client_secret
      when: "{{ vars['client_secret'] == 'test' }}"
```

I have retained the original `Set service principal parameter` code from `ansible-aks` to show the effect. There are two `debug` tasks, one with `lookup` which will transform the variable for the `when` and the second using only `vars` which will not transform the variable.

You will see that only `service_principal` is output by the debug statement, `client_secret` is skipped, if `vars` === `lookup` you would expect both to output a result.

The key point here is: `sp` will contain **untransformed** variables after the `set_fact` is executed with the original code. This results in incorrect untransformed variables being passed to Azure which in turn causes the call to fail. This is easy to see when `-vvv` is used.

Reply:
OK, I know what you mean but you are talking about a specific scenario that is not fully correct:

The Ansible way is this:

```yaml
---
- name: Lookup Vars
  hosts: localhost
  gather_facts: false
  connection: local

  vars:
    test_var: "test"
    service_principal: "{{ test_var }}"
    client_secret: "{{ test_var }}"

  tasks:

    - debug:
        var: service_principal
      when: "{{ lookup('vars', 'service_principal') == 'test' }}"

    - debug:
        var: client_secret
      when: client_secret == 'test'
```

That's why there are different rules for the when, to avoid this kind of output.

Reply:
@imjoseangel the example was designed to show the difference in behaviour between `vars` and `lookup`. Your example, while correct in terms of markup, does not demonstrate the problem the PR is addressing.

Reply:
Complete example:
```yaml
- hosts: localhost

  vars:
    my_var: test
    my_service_principal: "{{ my_var }}"
    my_client_secret: "{{ my_var }}"

  tasks:
      - include_role:
           name: azure.aks
        vars:
           name: akscluster
           resource_group: aksroletest
           location: eastus
           service_principal: "{{ my_service_principal }}"
           client_secret: "{{ my_client_secret }}"
```

Yields in the JSON:
```json
{
        "module_args": {
            "aad_profile": null,
            "ad_user": null,
            "addon": null,
            "adfs_authority_url": null,
            "agent_pool_profiles": [
                {
                    "count": 3,
                    "dns_prefix": null,
                    "name": "nodepool1",
                    "os_disk_size_gb": 30,
                    "os_type": "Linux",
                    "ports": null,
                    "storage_profiles": "ManagedDisks",
                    "vm_size": "Standard_DS1_v2",
                    "vnet_subnet_id": "/subscriptions/87970e72-3a66-4dd9-b051-1eba48a55882/resourceGroups/aksroletest/providers/Microsoft.Network/virtualNetworks/akscluster8f282/subnets/akscluster8f282"
                }
            ],
            "api_profile": "latest",
            "append_tags": true,
            "auth_source": null,
            "cert_validation_mode": null,
            "client_id": null,
            "cloud_environment": "AzureCloud",
            "dns_prefix": "akscluster",
            "enable_rbac": true,
            "kubernetes_version": "1.12.8",
            "linux_profile": {
                "admin_username": "azureuser",
                "ssh_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC79uN53bsaT4PuO6gyo+U8OPLpWo4EAwqkr5Amx80dQ7wyDDaByRubZ+Te0za+Koj9cN8pYirYSaCQvV2nRzbZ74u3ain4adwlnskVoQCf+sC8E1hcIT/bMTKLorQ6zHOVOFptuD+MVXoH6IVSS03nh+0d32imzjHfVBcblTMMGizhOBIbuR9wumNeZjvChgX5vFzai6/e2p3u9AXGVjae/h3mTCu9nmkVVCFGa0jogrZk26uHt2dbZ8WbYnu37WTqjqfdrQXZhJrULMN5PwvRzeJf7LGTPXngTGJR5VH/mdqpvQU8UAk888XMQqMsZRid+0Ggw2m1/MnQo5l6meA/ craig@debian-minimal"
            },
            "location": "eastus",
            "name": "akscluster",
            "network_profile": {
                "dns_service_ip": null,
                "docker_bridge_cidr": null,
                "network_plugin": "kubenet",
                "network_policy": null,
                "pod_cidr": null,
                "service_cidr": null
            },
            "password": null,
            "profile": null,
            "resource_group": "aksroletest",
            "secret": null,
            "service_principal": {
                "client_id": "{{ my_service_principal }}",
                "client_secret": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
            },
            "state": "present",
            "subscription_id": null,
            "tags": null,
            "tenant": null
        }
    },
    "msg": "Error creating the AKS instance: Operation failed with status: 'Bad Request'. Details: "
}
```

As you can see in the JSON the `service_principal.client_id` parameter is sent to Azure **untransformed**.

Reply:
Cool I understand. I have added your change in my PR. Probably you can open it and add the rest of the name changes you have in yours. What do you think?

Reply:
@imjoseangel
My changes are already in my fork here: https://github.com/craigminihan/ansible-aks/commits/master

It might be easier to let this PR go thru by itself then I can PR the rest after. Otherwise you can just rebase your branch from my master and the changes will appear in your PR. 

Might be best to let the project team decide which way to accept changes tho.

Reply:
Sure! Thanks for the comments and learnings!